### PR TITLE
Reserve memory for DICE data from Stage 0

### PR DESCRIPTION
**Pull request by conradgrobler**
We need to reserve the memory that Stage 0 uses to pass the DICE data to the next stage. This stops the kernel from using this memory, and allows the next stage to find the memory.

The logic for modifying the E820 table has also been extended to make reserving memory simpler.

Ref https://github.com/project-oak/oak/issues/4074

### CODE REVIEW

File: stage0/src/zero_page.rs [hunk 1]
```diff
@@ -188,6 +154,94 @@ impl ZeroPage {
...
+    /// address).
+    ///
+    /// If the new entry overlaps with one or more existing entries in the table, the effect depends
+    /// on the entry type:
+    ///
+    /// - If the new entry has the same type as the existing entry the entries will be merged into a
+    ///   single entry. Two adjacent entries of the same type will also be merged.
+    /// - If the new entry has a different type the new entry will be inserted as is and the
+    ///   existing entry will be modified (trimmed or split into two entries) to avoid the overlap.
+    pub fn insert_e820_entry(&mut self, entry: BootE820Entry) {
```
Comment (ipetr0v) [inline stage0/src/zero_page.rs:167]: Is current plan for passing DICE data to Layer1 to use E820 for both RKernel and Linux Kernel?

**Review by ipetr0v — APPROVED** [review]

**Review by andrisaar — APPROVED** [review]
I'll have to admit that I'm relying on your test coverage to make sure all the boundary conditions work properly :)
